### ISSUE TITLE
Provider has apparently gotten more unstable recently (seems to be related to the Azure Graph API)

### ISSUE THREAD

**Issue by ekristen** [hashicorp/terraform-provider-azurerm#17565]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.2.4

### AzureRM Provider Version

3.7.0

### Affected Resource(s)/Data Source(s)

azurerm_*

### Overview of Problem

This is a meta bug report as this is provider wide not just on specific resources. It is unclear if this is a result of changes on the Azure side or using the Graph API or a combination but this provider has become pretty unreliable failing frequently without retries.

Errors range from 'waiting for completion' to 'encountered retryable error' but ironically doesn't retry.

There are a number of issues open that reference these sort of errors already but as it pertains to a particular resources. 

I believe there needs to be some better internal retry logic instead of immediate fail upon a single error.

One of the biggest issues with the "something went wrong" retryable is it causes terraform to not record the resource in state even though it was reported as successfully created! The call after it errors and terraform exits.

The retryable error frequency can be reduced by forcing the parallelism below 5, like 3 or 4 almost makes the retryable errors from never occurring.

However errors like "error waiting for completion" still happens and forces you to have to run another plan and another apply to make it work.

### Debug Output/Panic Output

```shell
│ Error: waiting for completion of Security Group Association for Network Interface: (Name "test-azure-j0xyh-private-server" / Resource Group "test-azure-j0xyh"): Code="InternalServerError" Message="An error occurred." Details=[]
```
```


### Expected Behaviour

The provider should retry errors more times at a minimum especially ones that report themselves as being retryable.

### Actual Behaviour

The provider hits errors and just exist, doesn't seem to retry.

### Steps to Reproduce

_No response_

### Important Factoids

_No response_

### References

_No response_

**Comment by erikerikson** [comment]
@magodo you surely meant @ekristen.

**Comment by ekristen** [comment]
1. Perhaps RetryableError from Azure SDK can be added to the retry but on a provider wide scale. I will try and get more examples, I wasn't saving off the errors at first as I hadn't seen the pattern. As I run into it more I will share the results. I might be able to throw a test repo together too. The code I'm using can't directly be shared at the moment, but it's nothing to spectacular: virtual network, 3 subnets, 3 route tables, 1 nat gateway, 2 linux system. 

2. I don't think I agree with that logic. For example ...

```
Error: waiting  for completion of Security Group Association for Network Interface:  (Name "test-terraform-azure-j0xyh-private-server" / Resource Group  "test-terraform-azure-j0xyh"): Code="InternalServerError"  Message="An error occurred." Details=[]
--
```

Surely this sort of error should be retried?  My concern is that it hits this error and it's just an ALL stop on the entire terraform run. What's more is that all terraform logs before this error show that all security group assocations were created successfully.

3. This happens most frequently with the linux vm creation in my experience. Next time I see it I'll copy the logs over. Basically Terraform reports that the linux vm was created, then some random error like the one item 2 is thrown and terraform just exits and state isn't saved properly. If you run plan again, the provider will state that there's already a resource created that matches the request but it's not in state.

### Conclusion

This seems to be a provider wide azure graph error handling issue.

**Comment by tombuildsstuff** [comment]
@ekristen the Azure Networking API requires locking in various points when provisioning resources in parallel - whilst the provider handles this when provisioning through a single instance of the Provider, if you're running multiple Terraform deployments in parallel then the locks we use within the Provider won't apply. Are you able to confirm if you're running multiple Terraform deployments against the same Subscription/related resources (e.g. the NSG from the original error above)?

Whilst I'd agree that the base layer used by the SDK should account for `RetryableErrors` - that's something which'd be part of the meta issue tracked in #10858 - unfortunately the behaviour of the API requires that certain resources/changes are performed serially rather than in parallel, just retrying would ultimately end up exceeding the rate limit for some users (as we don't know _when_ it's safe to retry).

From what I can see here this'd be partly fixed by #10858 - but ultimately we're gated by the behaviour of the Azure API here (hence the locking to ensure resources are provisioned serially internally as required). Are you able to provide a reproducible Terraform configuration so that we can see if there's anything else we can do for this in the interim?

Thanks!

**Comment by ekristen** [comment]
@tombuildsstuff that's very interesting information. Thank you.

I may have run multiple terraform instances against the same subscription, but never the same resources. Recently during testing I've only been doing single terraform runs and run into many of these errors, turning parallelism down to 4 from the default 10 makes the "retryable errors" less frequent, but instead, I run into others. 

I would say I'm seeing errors 40% of the time on very very simple infrastructure builds, but I'll do some more tracking and testing and capture more output to be certain.

Agreed on RetryableErrors and glad it's being tracked already and I understand there's a problem with serial vs parallel with Azure :(

Let me see what I can do to put together a reproducible project for this and follow up.

**Comment by ekristen** [comment]
@tombuildsstuff as a follow up the same range that has the `Error: waiting  for completion of Security Group Association for Network Interface:` now cannot be destroyed because it is throwing the error below. Basically because it failed to do something with the network group previously, something was missed in state save, so now it can't cleanup properly.

This speaks to item 3 above.

```
│ Error: deleting Network Security Group "test-terraform-azure-j0xyh-private-server" (Resource Group "test-terraform-azure-j0xyh"): network.SecurityGroupsClient#Delete: Failure sending request: StatusCode=400 -- Original Error: Code="InUseNetworkSecurityGroupCannotBeDeleted" Message="Network security group /subscriptions/xxxxxxxx-yyyy-yyyy-yyyy-zzzzzzzzzzzzz/resourceGroups/test-terraform-azure-j0xyh/providers/Microsoft.Network/networkSecurityGroups/test-terraform-azure-j0xyh-private-server cannot be deleted because it is in use by the following resources: /subscriptions/xxxxxxxx-yyyy-yyyy-yyyy-zzzzzzzzzzzzz/resourceGroups/test-terraform-azure-j0xyh/providers/Microsoft.Network/networkInterfaces/test-terraform-azure-j0xyh-private-server. In order to delete the Network security group, remove the association with the resource(s). To learn how to do this, see aka.ms/deletensg." Details=[]
```

**Comment by tombuildsstuff** [comment]
@ekristen the "association" resources are virtual resources which mutate other resources, which require locks for coordination (else they'll knock each other out) - so if multiple are being deployed in parallel that'd explain this. Can you talk more about the deployment scenario here?

**Comment by ekristen** [comment]
@tombuildsstuff I haven't been able to reproduce all the errors but with this https://github.com/ekristen/bug-terraform-azurerm-lock -- I've been able to trigger a couple. 

It's possible the use of modules triggers other errors. In my other terraform I'm using modules, this is where I see the issues with the association resource errors more frequently, I'm looking at how I can put together a reproducable repo with all the modules I have in use. The modules simple isolate like resoruces together, IE one creates the netowrk, one for creating subnets (and their route tables), a module for creating a linux vm with a public ip and network configuration, network security group too.

**Comment by ekristen** [comment]
@tombuildsstuff here is another issue ... this happened to day where a VM was destroyed and created and azurerm encountered an internal error from the API and didn't record the VM in state, but the VM exists and now exists outside the terraform state.

```module.windows.azurerm_windows_virtual_machine.default: Still destroying... [id=/subscriptions/xxxxx...ines/tpl-terraform-azure-ekgpc-windows, 10s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still destroying... [id=/subscriptions/xxxxx...ines/tpl-terraform-azure-ekgpc-windows, 20s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still destroying... [id=/subscriptions/xxxxx...ines/tpl-terraform-azure-ekgpc-windows, 30s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still destroying... [id=/subscriptions/xxxxx...ines/tpl-terraform-azure-ekgpc-windows, 40s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Destruction complete after 49s
module.windows.azurerm_windows_virtual_machine.default: Creating...
module.windows.azurerm_windows_virtual_machine.default: Still creating... [10s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [20s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [30s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [40s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [50s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [1m0s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [1m10s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [1m20s elapsed]
module.windows.azurerm_windows_virtual_machine.default: Still creating... [1m30s elapsed]
╷
│ Error: waiting for creation of Windows Virtual Machine: (Name "tpl-terraform-azure-ekgpc-windows" / Resource Group "tpl-terraform-azure-ekgpc"): Code="OSProvisioningInternalError" Message="OS Provisioning failed for VM 'tpl-terraform-azure-ekgpc-windows' due to an internal error."
│
│   with module.windows.azurerm_windows_virtual_machine.default,
│   on ../tf-azure-windows/main.tf line 148, in resource "azurerm_windows_virtual_machine" "default":
│  148: resource "azurerm_windows_virtual_machine" "default" {
│
╵```

**Comment by ekristen** [comment]
I can reproduce this almost 100% of the time. I haven't quite identified why yet, but my code definitely does.

**Comment by tombuildsstuff** [comment]
Unfortunately these are all Upstream API issues rather than something Terraform can do something about, whilst we could (for example) look to delete the VM and recreate it if it fails provisioning, the logic around when that's safe to do (even if we only did it at creation, for example) is pretty non-trivial (since Azure uses the `name` as the unique identifier, not all Azure Services support reusing the name straight away, some newer resources now block reusing the names too) - so I think this has to be fixed on the API end.

Have you opened a support ticket with Microsoft to dig in on their end?

**Comment by ekristen** [comment]
I understand the challenges with the Azure API, it's very frustrating and often counter intuitive to use, not to mention the instability. I have not opened an issue with Microsoft. Unfortunately my track record with opening issues with Microsoft and getting valuable responses is not very good.

With that in mind, I still believe there's a provider issue here. In the case of the windows vm creation failure, it's being created, it's just that whatever call the provider is making during the "still creating..." is throwing an "internal error" and immediately exiting and dropping state. This is bad in my opinion. 

The only scenario I can think of that I would want the "still creating..." to just exit is perhaps after it's timeout, any errors should be retried, and/or drop a message saying still creating (received error blah, 1 of 10 allowed) and retry again. All efforts should be taken to not drop state.

Thanks

**Comment by tombuildsstuff** [comment]
@ekristen 

>  it's just that whatever call the provider is making during the "still creating..." is throwing an "internal error" and immediately exiting and dropping state

Unfortunately the Azure API differs in it's behaviour depending on the API (and Payload) being used - in some APIs the resource can be marked as Failed and eventually provisioned, in others it can be Failed and remain in a Failed (unusable) state - in others it can be Failed and rolled back (e.g. removed) - so unfortunately we don't have enough context on what the Azure API's going to do to these resources when they fail to provide guarantees here.

Specifically within Terraform we don't mark a resource as successfully provisioned until it's in a Ready state (for example, for a VM that it's fully available not just starting the boot process - or for KeyVault that the DNS has finished replicating). Since we don't have these guarantees when the resource is marked as Failed, unfortunately we can't guarantee the state of the resource, and so we can't mark this as successfully provisioned (e.g. track this in Terraform) since it's not necessarily correct.

As mentioned above unfortunately the specific bugs in question need to be fixed by Azure in the Azure API, due to the aforementioned limitations unfortunately there's not much we can do to workaround this given the lack of information within the Azure API, without as @magodo has mentioned, polling indefinitely on something that'll never recover.

I appreciate that's probably not the answer you're looking for, but given these issues need to be fixed by the Azure Service Team I'd recommend opening an issue with Microsoft Support, who should be able to get the Service Teams to look into the Internal Server Errors being returned here - since they'd need to correlate these on the backend to be able to diagnose the issue within their API; however since unfortunately there's not much we can do to workaround these InternalServerErrors being returned from the APIs in the interim, I'm going to close this issue for the moment.

Thanks!

**Comment by ekristen** [comment]
Hey @tombuildsstuff I appreciate your position, you have some valid points, but so do I. I don't think closing this issue out is the right move.

> Unfortunately the Azure API differs in it's behavior depending on the API (and Payload) being used - in some APIs the resource can be marked as Failed and eventually provisioned, in others it can be Failed and remain in a Failed (unusable) state - in others it can be Failed and rolled back (e.g. removed) - so unfortunately we don't have enough context on what the Azure API's going to do to these resources when they fail to provide guarantees here.

I understand, but the provider has to make best effort to make the behavior consistent and behavior I mean how the provider handles these edge cases such that the end user experience is consistent.

> Specifically within Terraform we don't mark a resource as successfully provisioned until it's in a Ready state (for example, for a VM that it's fully available not just starting the boot process - or for KeyVault that the DNS has finished replicating). Since we don't have these guarantees when the resource is marked as Failed, unfortunately we can't guarantee the state of the resource, and so we can't mark this as successfully provisioned (e.g. track this in Terraform) since it's not necessarily correct.

I would argue that if the resource exists, regardless of state that it should be recorded into state, understanding that you are going for "Ready" this is where the timeout of the resource provisioning comes in within the terraform provider (I know the AWS provider does this internally). For various services it was X minutes before bailing, but when it bails it still **records** it into state.

> As mentioned above unfortunately the specific bugs in question need to be fixed by Azure in the Azure API, due to the aforementioned limitations unfortunately there's not much we can do to workaround this given the lack of information within the Azure API, without as @magodo has mentioned, polling indefinitely on something that'll never recover.

I agree there are problems, but there are problems with the provider as well. I would never suggest polling indefinitely. Take AWS provider for example, there are built in timeouts such that if a resource doesn't reach a specific state by specific time period it will exit, **however** the biggest difference is that it still records the resource in the state.

It's a tough situation but given we know there are problems with Azure, I would say that all efforts should be taken to retry until a retry or time threshold is hit, record what is known and exit. Upon secondary runs it recognizes the resource in state and see's that it's not in the correct and expected state and proceeds to try and resolve it.

---

Ultimately I agree with you the Azure API needs work. If you'd prefer me to open separate issues, I'm happy to do that, as I still feel there's room for improvement in the provider.

I've not had good luck reaching out to Microsoft for support on a number of simple topics, getting them to acknowledge problems with their API is highly unlikely.

It would be more beneficial to the terraform community to help guard against failure points of the Azure API in the provider.

**Comment by ekristen** [comment]
I provided a more specific example, but this appears to go beyond just one resource, I think this issue is still valid for review.
